### PR TITLE
Store Orders: Add state to track edits made to orders

### PR DESCRIPTION
**Pull request by ryelle**
This PR adds a reducer and actions to track order edits. This adds a subtree to `extensions.woocommerce.ui.orders`:

```
edits: {
	currentlyEditingId: {},
	changes: {
		... order data ...
	},
},
```

This tracks one currently-editing order, which can be an existing order, or a newly created one. Newly created orders get placeholder IDs in the format `{ placeholder: 'order_1' }`. This assumes we're only editing one order at a time (a true limitation of the current UI, bulk management will happen eventually, and this can be updated then).

**To test:**

There are no functionality changes yet, so this can only be tested with the code tests:

- Run `npm run test-client client/extensions/woocommerce/state/ui/orders`

### CODE REVIEW

File: client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+		expect( newState ).to.eql( {
+			currentlyEditingId: 42,
+			changes: {
+				id: 42,
+				first_name: 'Fiona',
+			},
+		} );
+	} );
+
+	it( 'should store a generated an ID for a created order', () => {
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js]: minor: `an` here can be dropped.

**Comment by ryelle** [comment]
@justinshreve I've had to make a change to the reducer to support the nested object structure in the orders API format (like `billing`, `shipping`, etc). Now the reducer merges the action's order into the state. Can you check over 160b73a?

File: client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+			order: {
+				id: 40,
+				billing: {
+					first_name: 'Joan',
+					last_name: 'Watson',
+				}
+			},
+		};
+		const originalState = deepFreeze( {
+			currentlyEditingId: 40,
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js]: Should currentlyEditingId be null instead to best exercise the reducer for action type WOOCOMMERCE_UI_ORDERS_EDIT?

File: client/extensions/woocommerce/state/ui/orders/edits/reducer.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function currentlyEditingId( state = null, action ) {
+	switch ( action.type ) {
+		case WOOCOMMERCE_UI_ORDERS_EDIT:
+			if ( action.order && action.order.id ) {
+				return action.order.id;
+			}
+			return { placeholder: uniqueId( 'order_' ) };
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/orders/edits/reducer.js]: If I make two edits to a not-yet-persisted-to-the-server order, won't this placeholder ID roll? That seems bad. Maybe you need a conditional to test for placeholder before rolling it?

File: client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			changes: {
+				billing: {
+					first_name: 'Joan',
+					last_name: 'Watson',
+				}
+			},
+		} );
+	} );
+
+	it( 'should merge updates to an order if new fields are passed in', () => {
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js]: Probably need to add a similar test for not-yet-persisted-to-the-server orders. Please consider comment above about whether placeholder ID should be rolling or not (that might be a good thing to test itself.)

File: client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js (hunk 2 above)
Comment (ryelle) [reply client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js]: The test above this one checks that `currentlyEditingId` is correctly set from the empty state - and it doesn't make sense that there would be existing edits in the state without having `currentlyEditingId` defined.

File: client/extensions/woocommerce/state/ui/orders/edits/reducer.js (hunk 3 above)
Comment (ryelle) [reply client/extensions/woocommerce/state/ui/orders/edits/reducer.js]: As soon as a change is made to the order it gets its placeholder ID, so following updates will pass back the ID (as of https://github.com/Automattic/wp-calypso/pull/17966/commits/6b5c74b690c2ccc5b5b27f0628cad3b88ac213c5, in the selectors PR #17966), and it shouldn't update.

File: client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js (hunk 4 above)
Comment (ryelle) [reply client/extensions/woocommerce/state/ui/orders/edits/test/reducer.js]: Oh I see now - the placeholder ID is passed back to the reducer via the action, so there should be no difference between newly created or edited orders, but I've added the test anyway.

**Pull request merged**